### DISCUSSION TITLE
confusing behaviors of `dd` commands on compile-time

### DISCUSSION THREAD

**Discussion by Icemic** [immortalwrt/immortalwrt#629] [General]
Hi, I'm trying to compile a custom-sized image recently and the some `dd` commands during rootfs image generating confused me.

setting kernel partition size to 128M and root partition size to 4096M,

when generating squashfs images it performs like

```shell
+ dd if=/dev/zero of=/app/immortalwrt/build_dir/target-x86_64_musl/linux-x86_64/tmp/immortalwrt-x86-64-generic-squashfs-combined-efi.img bs=512 seek=263168 conv=notrunc count=8388608
197496+1 records in
197496+1 records out
101118353 bytes (101 MB, 96 MiB) copied, 0.418215 s, 242 MB/s
dd if=/app/immortalwrt/build_dir/target-x86_64_musl/linux-x86_64/tmp/immortalwrt-x86-64-generic-squashfs-rootfs.img of=/app/immortalwrt/build_dir/target-x86_64_musl/linux-x86_64/tmp/immortalwrt-x86-64-generic-squashfs-rootfs.img.new bs=4294967296 conv=sync
0+1 records in
1+0 records out
4294967296 bytes (4.3 GB, 4.0 GiB) copied, 15.3191 s, 280 MB/s
mv /app/immortalwrt/build_dir/target-x86_64_musl/linux-x86_64/tmp/immortalwrt-x86-64-generic-squashfs-rootfs.img.new /app/immortalwrt/build_dir/target-x86_64_musl/linux-x86_64/tmp/immortalwrt-x86-64-generic-squashfs-rootfs.img
8388608+0 records in
8388608+0 records out
4294967296 bytes (4.3 GB, 4.0 GiB) copied, 19.7148 s, 218 MB/s
```

**So my questions are:**

1. why `bs` is set so large for rootfs.img which seems always to be equal to the image size? (while `bs` is set fixed to `512` for combined one). This however results in any image that is larger than the memory size cannot be generated.
2. why there is a `dd if=* if=*.new` which appears to be useless?

**what is more confusing** is when I tried to generate ext4 images, rootfs generation logs became
```shell
dd if=/app/immortalwrt/build_dir/target-x86_64_musl/linux-x86_64/tmp/immortalwrt-x86-64-generic-ext4-rootfs.img of=/app/immortalwrt/build_dir/target-x86_64_musl/linux-x86_64/tmp/immortalwrt-x86-64-generic-ext4-rootfs.img.new bs=4294967296 conv=sync
0+3 records in
3+0 records out
```

which makes rootfs.img three times larger than combined.img nearly. I've try to use a smaller root size, for example, 512M, but it just changed to be `0+2 records in`.

Not sure if it's a bug because I could not find the code which executes those `dd` commands.

**Comment by Icemic** [comment]
In addition, all those compiling is under docker image `ubuntu:18.04` with non-root user.

`.config` file: https://gist.github.com/Icemic/ded409759d5834ee35b7d58f300f2bed

**Comment by Icemic** [comment]
或者告诉我这段代码在哪里，也许我能尝试解决掉这个问题

**Comment by 1715173329** [comment]
This is defined by upstream OpenWrt, as it just works so I never do a research on it.

Related commands can be found there:
`include/image.mk`
`include/image-commands.mk`
